### PR TITLE
[ecosystem] [1/3] add token tables to indexer

### PR DESCRIPTION
**Pull request by areshand**
## Motivation

To provide basic DB tables for token indexer API.

1. add models for serde token, token id, tokenData and token events
2. add a token processor to add records to token table

document the todos in https://github.com/aptos-labs/aptos-core/issues/1132 and will continuous to get the rest of items done. 

### Have you read the [Contributing Guidelines on pull requests](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

test run token processor with fullnode and see the table created as expect.

Screen Shot 2022-05-22 at 3 12 12 PM

## Related PRs

(If this PR adds or changes functionality, please take some time to update the docs at https://github.com/aptos-labs/aptos-core/tree/main/developer-docs-site, and link to your PR here.)

### CODE REVIEW

File: ecosystem/indexer/src/models/token.rs [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+#[primary_key(token_id)]
+pub struct Token {
+    pub token_id: String,
+    pub creator: String,
+    pub collection: String,
+    pub name: String,
+    pub description: String,
+    pub max_amount: Option<i64>,
+    pub supply: i64,
+    pub uri: String,
```
Comment (hariria) [inline ecosystem/indexer/src/models/token.rs:21]: This is the arweave / IPFS URI correct?

File: ecosystem/indexer/src/models/token.rs [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+    pub minted_at: chrono::NaiveDateTime,
+
```
Comment (hariria) [inline ecosystem/indexer/src/models/token.rs:23]: Nit remove line

File: ecosystem/indexer/src/models/token.rs [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+use crate::{models::events::Event, schema::tokens};
+use aptos_rest_client::types;
+use std::{collections::HashMap, fmt, fmt::Formatter, str::FromStr};
+
+use serde::{Deserialize, Deserializer, Serialize};
+
+#[derive(Associations, Debug, Identifiable, Insertable, Queryable, Serialize)]
+#[diesel(table_name = "tokens")]
+#[primary_key(token_id)]
+pub struct Token {
```
Comment (hariria) [inline ecosystem/indexer/src/models/token.rs:13]: Can we add `Clone` macros to these structs as well? If not, no worries

File: ecosystem/indexer/src/models/token.rs [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+    pub max_amount: Option<i64>,
+    pub supply: i64,
+    pub uri: String,
+    pub minted_at: chrono::NaiveDateTime,
+
+    pub inserted_at: chrono::NaiveDateTime,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct TokenId {
```
Comment (hariria) [inline ecosystem/indexer/src/models/token.rs:28]: Can we add `Clone` macros to these structs as well?

File: ecosystem/indexer/src/models/token.rs [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+}
+
+impl fmt::Display for TokenId {
+    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
+        write!(f, "{}::{}::{}", self.creator, self.collection, self.name)
+    }
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct TokenData {
```
Comment (hariria) [inline ecosystem/indexer/src/models/token.rs:47]: Can we add `Clone` macros to these structs as well?

File: ecosystem/indexer/src/models/token.rs [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+    pub collection: String,
+    pub description: String,
+    pub name: String,
+    #[serde(deserialize_with = "deserialize_option_from_string")]
+    pub maximum: MoveOption<i64>,
+    #[serde(deserialize_with = "deserialize_option_from_string")]
+    pub supply: MoveOption<i64>,
+    pub uri: String,
```
Comment (hariria) [inline ecosystem/indexer/src/models/token.rs:55]: Is this the actual image uri or the arweave uri to the json? If its the actual image uri can we make the field name `image`?

File: ecosystem/indexer/src/models/token.rs [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct TokenData {
+    pub collection: String,
+    pub description: String,
+    pub name: String,
+    #[serde(deserialize_with = "deserialize_option_from_string")]
+    pub maximum: MoveOption<i64>,
+    #[serde(deserialize_with = "deserialize_option_from_string")]
+    pub supply: MoveOption<i64>,
```
Comment (hariria) [inline ecosystem/indexer/src/models/token.rs:54]: I dont think this should be an option? It's not nullable in schema.rs

File: ecosystem/indexer/src/models/token.rs [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+                Ok("".parse::<T>().unwrap())
+            } else {
+                v["vec"][0].parse::<T>().map_err(D::Error::custom)
+            }
+        },
+    )
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct WithdrawEventType {
```
Comment (hariria) [inline ecosystem/indexer/src/models/token.rs:103]: Can we add `Clone` macros to these structs as well?

File: ecosystem/indexer/src/models/token.rs [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+    #[serde(deserialize_with = "types::deserialize_from_string")]
+    pub amount: i64,
+    pub id: TokenId,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
```
Comment (hariria) [inline ecosystem/indexer/src/models/token.rs:109]: Can we add `Clone` macros to these structs as well?

File: ecosystem/indexer/src/models/token.rs [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+pub struct DepositEventType {
+    #[serde(deserialize_with = "types::deserialize_from_string")]
+    pub amount: i64,
+    pub id: TokenId,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct CreationEventType {
```
Comment (hariria) [inline ecosystem/indexer/src/models/token.rs:117]: Can we add `Clone` macros to these structs as well?

File: ecosystem/indexer/src/models/token.rs [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+    pub id: TokenId,
+    pub token_data: TokenData,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct MintEventType {
```
Comment (hariria) [inline ecosystem/indexer/src/models/token.rs:123]: Can we add `Clone` macros to these structs as well?

File: ecosystem/indexer/src/models/token.rs [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+    DepositEvent(DepositEventType),
+    CreationEvent(CreationEventType),
+    MintEvent(MintEventType),
+    BurnEvent,
+}
+
+impl TokenEvent {
+    pub fn from_event(event: &Event) -> Option<TokenEvent> {
+        let data = event.data.clone();
+        match event.type_.as_str() {
```
Comment (hariria) [inline ecosystem/indexer/src/models/token.rs:141]: nice

File: ecosystem/indexer/src/models/token.rs (hunk 1 above)
Comment (areshand) [reply ecosystem/indexer/src/models/token.rs:21]: Yes, this is a URI that we can use for either arweave  or IPFS

File: ecosystem/indexer/src/models/token.rs (hunk 3 above)
Comment (areshand) [reply ecosystem/indexer/src/models/token.rs:13]: added

File: ecosystem/indexer/src/models/token.rs (hunk 4 above)
Comment (areshand) [reply ecosystem/indexer/src/models/token.rs:27]: added

File: ecosystem/indexer/src/models/token.rs (hunk 6 above)
Comment (areshand) [reply ecosystem/indexer/src/models/token.rs:54]: this is the metadata uri to the json.
I will create another metadata table, which we can have a field called URL?

Comment (hariria) [reply ecosystem/indexer/src/models/token.rs:54]: Yeah, maybe make a table called `NFTMetadata` or something like that (other tokens will have different metadata schemas)

File: ecosystem/indexer/src/models/token.rs (hunk 7 above)
Comment (areshand) [reply ecosystem/indexer/src/models/token.rs:53]: These two fields are mapped the option fields in the move `Tokendata` definition. We can directly deserialize from the tokenData json string from raw event. Later, to store the data in DB, I transform the `TokenData` here to match the Token defined at line 13, where `supply` in not optional and `max_amount` is still optional. In schema.rs, the `max_amount` is also nullable as users may not specify the max when creating the token.

File: ecosystem/indexer/src/models/token.rs (hunk 9 above)
Comment (areshand) [reply ecosystem/indexer/src/models/token.rs:108]: added

File: ecosystem/indexer/src/models/token.rs (hunk 10 above)
Comment (areshand) [reply ecosystem/indexer/src/models/token.rs:116]: added

File: ecosystem/indexer/src/models/token.rs (hunk 6 above)
Comment (areshand) [reply ecosystem/indexer/src/models/token.rs:54]: sure

**Review by hariria — APPROVED** [review]
Once the `NFTMetadata` table is added, it LGTM

**Pull request merged**